### PR TITLE
Remove the div wrapper from the aligned image blocks

### PR DESCRIPTION
**Pull request by youknowriad**
Related #38613 

For historic reasons, the image block had a div wrapper around the block when it was left/right or center aligned (Frontend only). This has a number of issues:

 - It's inconsistent with other blocks,
 - It makes the markup of the frontend and the backend different

In this PR, I'm updating the image block to remove that extra div and make the block work like any other block. That said, and since this block has been like that for a long time, I'm dynamically restoring that div in the frontend for classic themes (themes without theme.json and layout support). The idea is that if you don't use the layout, you're providing your own alignment styles, so we should limit the impact with your theme's existing styles for backward compatibility. 

**Testing instructions**

This is a somewhat impactful change that we need to check properly:

 - We need to test that image blocks (aligned left, right or center) with the old markup are upgraded properly to the new markup without block invalidation when opening the editor.
 - We need to check that the markup in the frontend stayed exactly the same for classic themes for these blocks (avoid breakage)
 - We need to check that for FSE themes, there's no extra div in the frontend for aligned (left, right or center) blocks making the behavior consistent between the frontend and the backend.

I'd appreciate a lot of help testing this PR cause it can be impactful.

### CODE REVIEW

File: lib/block-supports/layout.php
Comment (youknowriad) [inline lib/block-supports/layout.php:248]: @ockham @mcsf if you have any thoughts about my regex (I'm not an expert there :P)

**Comment by jasmussen** [comment]
From a first glance this works really well, with editor and frontend looking right:
Screenshot 2022-02-09 at 10 46 06

Screenshot 2022-02-09 at 10 46 18

The containing div remains in the editor, and is likely responsible for the margin inconsistency. This is also likely what https://github.com/WordPress/gutenberg/pull/38613 will fix as a part 2. But on the frontend, things are looking beautiful:
Screenshot 2022-02-09 at 10 47 55

In trunk, things look largely the same in the editor, but a big difference can be seen in TT2, where this patch fixes an issue without the theme having to do anything at all. Here's trunk TT2 frontend:
Screenshot 2022-02-09 at 10 54 25

Now there's an argument to be made that containing floats in the main column is _desirable_. But that's [still possible with the div-less markup](https://codepen.io/joen/pen/zLWvrW), and for block themes it's arguably more important that editor matches frontend, which this PR addresses.

**Comment by youknowriad** [comment]
> The containing div remains in the editor, and is likely responsible for the margin inconsistency

Yes, that's the role of the other PR https://github.com/WordPress/gutenberg/pull/38613 which hopefully we can get land right after we merge this one (if we merge this one)

**Comment by scruffian** [comment]
This looks great with emptytheme:
Screenshot 2022-02-09 at 10 10 01
Screenshot 2022-02-09 at 10 09 53

In twentytwentytwo things are different between the back and front end, but I think that's because of alignment styles in the theme.

**Review by scruffian — APPROVED** [review]

**Review by Mamaduka — COMMENTED** [review]
✅  Markup is updated correctly.
✅  Markup restoration for Classic Themes works as well.
✅  No extra `div` on the front-end for themes with `theme.json`.

I agree with Riad that more testing would be needed to land this change.

We should probably post a dev note for this change once both PRs are merged.

File: lib/block-supports/layout.php
Comment (Mamaduka) [reply lib/block-supports/layout.php:248]: **nitpick (non-blocker):**  We can probably return earlier if the block isn't an image or doesn't have `theme.json`.

```php
if (
  'core/image' !== $block['blockName'] ||
  WP_Theme_JSON_Resolver::theme_has_support()
) {
  return $block_content;
}
```

P.S. Sorry, I'm also not an expert here.

**Comment by kjellr** [comment]
> In twentytwentytwo things are different between the back and front end, but I think that's because of alignment styles in the theme.

If that's the case (and this PR gets merged), would someone mind opening a Trac issue to ensure this gets followed up on in Twenty Twenty-Two?

**Comment by youknowriad** [comment]
I'm not sure I see anything different between backend and frontend in 2022 (aside the zeroed margins which is fixed in the linked PR), @scruffian do you have some example content to test with?

**Comment by Mamaduka** [comment]
@youknowriad, I used this example by @MaggieCabrera for testing - https://gist.github.com/MaggieCabrera/799d09c819616354a1b8eb3605f4fc69.

**Comment by youknowriad** [comment]
@Mamaduka In that markup I don't see any image aligned left/right or center, meaning the current PR shouldn't have any impact there.

**Comment by Mamaduka** [comment]
I changed alignment for a few images, just for this PR 😄

**Comment by scruffian** [comment]
> I'm not sure I see anything different between backend and frontend in 2022 (aside the zeroed margins which is fixed in the linked PR), @scruffian do you have some example content to test with?

I've been doing some more testing and I can't replicate the issue now. I think I must have had a bad build before.

**Comment by youknowriad** [comment]
Ok I'm considering merging this PR, let me know if you have any reservations.

**Comment by jasmussen** [comment]
We know from past experience that in the cases where we can keep the markup the same, there are many benefits to be had. Especially for base ingredients like an image, there's all the more value to the simplicity this can bring. At this moment, block themes offer us a way to make the change happen. Merging it sooner can help us unearth side effects.

File: lib/block-supports/layout.php
Comment (youknowriad) [reply lib/block-supports/layout.php:248]: yeah, that's what it does already but since this is just an assignment, I think it's fine. (the variable is also used for the early return)

**Comment by ellatrix** [comment]
@youknowriad Does this allow us to remove the extra wrapper from the editor as well? As far as I remember, this was specifically there for image block compatibility.

https://github.com/WordPress/gutenberg/blob/cab8cc52cf715722d8d8aa081ee66744e991ab27/packages/block-editor/src/components/block-list/block.js#L122-L135

**Comment by Mamaduka** [comment]
@ellatrix, that will be handled by #38613.

**Comment by jasmussen** [comment]
> As far as I remember, this was specifically there for image block compatibility.

The comment about the main column, that should still be [possible to achieve](https://codepen.io/joen/pen/zLWvrW) with the wrapping `figure` element.

File: lib/block-supports/layout.php
Comment (mcsf) [reply lib/block-supports/layout.php:248]: Notes from reviewing the regular expression:

1. Word boundary needed before "wp-block-image", otherwise you would match `<figure class="not-wp-block-image […]`. **Fix:** `\bwp-block-image`.

2. The RE doesn't target the `class` attribute specifically; shouldn't it? Right now I can match `<figure not-class="...`.

3. No parens needed around each string: `(alignleft|alignright|aligncenter)`

4. Many of the remaining groups don't need capturing: we only care about `$match[1]` and `$match[2]`, so the above example could become `(?:alignleft|alignright:aligncenter)`, and same for `(?:\s|")`.

5. `((.|\S|\s)*)` is amusing! `\S|\s` means "I want to match something as well as its opposite", i.e. "everything", which is the same as `.`. So, unless I missed something, the outer group could be `.*`

6. `(\s|")`: I'm not sure what purpose this serves. Oh, maybe it's used as a word boundary after the alignment class? Right, and now I notice that there is missing word boundary at the start of the alignment class. You'll see that right now you're able to (incorrectly) match `zzzzalignleft`. So let's use proper word boundaries on both sides. **Fix:** `\b(?:alignleft|alignright|aligncenter)\b`.

7. (I don't think I'd ever seen the U (ungreedy) flag in use, that's cool)

Based on all of the above except for the fact that we don't target the `class` attribute, the RE could be:

```
/(^\s*<figure\b[^>]*)\bwp-block-image\b([^"]*\b(?:alignleft|alignright|aligncenter)\b[^>]*>.*<\/figure>)/U
```

I've tested this in the venerable [regexr.com](https://regexr.com), but not with live content, so take my review accordingly. :)

Comment (youknowriad) [reply lib/block-supports/layout.php:248]: For the first remark, it could also be "wp-block-image, is " a word boundary?

File: lib/block-supports/layout.php
Comment (mcsf) [inline lib/block-supports/layout.php:260]: What's the effect of the `static` keyword here? I thought static anonymous function declarations only mattered in the context of a class ([docs](https://www.php.net/manual/en/functions.anonymous.php#functions.anonymous-functions.static)).

File: lib/block-supports/layout.php
Comment (youknowriad) [reply lib/block-supports/layout.php:248]: 2- It could target class to be 100% accurate but I didn't think it was mandatory (like group blocks).

> 5- ((.|\S|\s)*) is amusing! \S|\s means "I want to match something as well as its opposite", i.e. "everything", which is the same as .. So, unless I missed something, the outer group could be .*

Indeed, but for some reason in the site I was using `.` was not catching everything (I think related to multiline)

6- `(\s|")` this one targets a space or end of attribute (if it's the last class)

File: lib/block-supports/layout.php
Comment (youknowriad) [reply lib/block-supports/layout.php:260]: I don't know it was copied from the "group" implementation but If I'm not wrong most of these `static` were introduced in a patch to support php 8 or something.

File: lib/block-supports/layout.php
Comment (mcsf) [reply lib/block-supports/layout.php:248]: > For the first remark, it could also be "wp-block-image, is " a word boundary?

That should work just as well, if you're sure that `wp-block-image` is always the first class.

> Indeed, but for some reason in the site I was using . was not catching everything (I think related to multiline)

Hm, interesting. Take my advice with caution, then. :) But make sure the trick is really necessary.

File: lib/block-supports/layout.php
Comment (ocean90) [inline lib/block-supports/layout.php:268]: Was there a reason for not using `render_block_core/image` as the filter?

Comment (youknowriad) [reply lib/block-supports/layout.php:268]: No particular reason, I guess my reasoning is that this change is tied to the "layout support" so I put here similarly to the previous one we have for group blocks. (same use-case)

Comment (youknowriad) [reply lib/block-supports/layout.php:268]: Or maybe you mean why not use the `render_block_core/image` filter name? If that's the case, I just didn't know that these block specific filters exist.

Comment (ocean90) [reply lib/block-supports/layout.php:268]: Yep, sorry for not being clear here. The filter was introduced in 5.7, see https://core.trac.wordpress.org/changeset/50123. It basically allows you to skip the `'core/image' !== $block['blockName']` check and only runs if the block is actually used on a page.

Comment (youknowriad) [reply lib/block-supports/layout.php:268]: Nice 👍 We should probably use it for both image and group filters.

Comment (youknowriad) [reply lib/block-supports/layout.php:268]: Opened a follow-up here https://github.com/WordPress/gutenberg/pull/38657

**Pull request merged**
